allennlp/data/dataset_readers/coreference_resolution/conll.py

```diff
@@ -0,0 +1,293 @@
+from allennlp.common.file_utils import cached_path
+from allennlp.data.dataset import Dataset
+from allennlp.data.dataset_readers.dataset_reader import DatasetReader
+from allennlp.data.fields import Field, ListField, TextField, IndexField, MetadataField, SequenceLabelField
+from allennlp.data.instance import Instance
+from allennlp.data.tokenizers import Token
+from allennlp.data.token_indexers import SingleIdTokenIndexer, TokenIndexer
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+class _DocumentState:
+    """
+    Represents the state of a document. Words are collected in a sentence buffer,
+    which are incrementally collected in a list when sentences end, representing
+    all tokens in the document.
+
+    Additionally, this class contains a per-id stacks to hold the start indices of
+    active spans (spans which we are inside of when processing a given word). Spans
+    with the same id can be nested, which is why we collect these opening spans
+    on a stack, e.g:
+
+     [Greg, the baker who referred to [himself]_ID1 as 'the bread man']_ID1
+
+    Once an active span is closed, the span is added to the cluster for the given id.
+    """
+    def __init__(self) -> None:
+        self.sentence_buffer: List[str] = []
+        self.sentences: List[List[str]] = []
+        self.num_total_words = 0
+        # Cluster id -> List of (start_index, end_index) spans.
+        self.clusters: DefaultDict[int, List[Tuple[int, int]]] = collections.defaultdict(list)
+        # Cluster id -> List of start_indices which are open for this id.
+        self.coref_stacks: DefaultDict[int, List[int]] = collections.defaultdict(list)
+
+    def assert_document_is_finished(self) -> None:
+        if self.sentence_buffer:
+            raise ConfigurationError("Processing was attempted on a document which had "
+                                     "incomplete sentences. Current sentence buffer "
+                                     "contains: {}".format(self.sentence_buffer))
+
+        if not self.sentences or self.num_total_words == 0:
+            raise ConfigurationError("Processing was attempted on a document with no sentences or words.")
+
+        if any(x for x in self.coref_stacks.values()):
+            raise ConfigurationError("Processing was attempted on a document which contains incomplete "
+                                     "(unclosed) spans within the scope of the document. Current ids "
+                                     "and start indices" " of unclosed spans: {}".format(self.coref_stacks))
+
+    def complete_sentence(self) -> None:
+        self.sentences.append(self.sentence_buffer)
+        self.sentence_buffer = []
+
+    def add_word(self, word) -> None:
+        self.sentence_buffer.append(word)
+        self.num_total_words += 1
+
+    def canonicalize_clusters(self) -> List[List[Tuple[int, int]]]:
+        """
+        The CONLL 2012 data includes 2 annotations which have identical
```

Comment:
It took me a while to understand this doc string.  I think it means "Sometimes the CONLL 2012 data includes 2 annotations with identical spans but different cluster ids.  When this happens, we want to merge the two clusters together.  This function merges clusters when they have the exact same mention."

Reply:
re-worded, thanks.